### ISSUE TITLE
Potential bug in hash key calculation

### ISSUE THREAD

**Issue by daniel-rikowski** [Shopify/bootsnap#318]
**tl;dr:** Hash calculation in the C extension might access invalid memory.

As I was working on #316 I noticed a potential bug in the calculation of hash keys which might cause segfaults or other unexpected behaviour.

The hash calculation done in `fnv1a_64` boils down to a loop over a string:

https://github.com/Shopify/bootsnap/blob/e8aad0ce200e05c935e0952aadc2e8a5e95e13f4/ext/bootsnap/bootsnap.c#L190-L193

In other words, it assumes that the string is null terminated.

This hash function is called from multiple places:

https://github.com/Shopify/bootsnap/blob/e8aad0ce200e05c935e0952aadc2e8a5e95e13f4/ext/bootsnap/bootsnap.c#L220
https://github.com/Shopify/bootsnap/blob/e8aad0ce200e05c935e0952aadc2e8a5e95e13f4/ext/bootsnap/bootsnap.c#L240

And one time indirectly through `bs_cache_path`
https://github.com/Shopify/bootsnap/blob/e8aad0ce200e05c935e0952aadc2e8a5e95e13f4/ext/bootsnap/bootsnap.c#L315-L320

**Now the problem:** `RSTRING_PTR` returns a pointer to whatever data structure is containing the string and it is not guaranteed that it is actually a null-terminated string. ([Documentation](https://docs.ruby-lang.org/en/2.7.0/extension_rdoc.html), search for `RSTRING_PTR`)

Look at the declaration of the internal data structure , depending on the length of the string Ruby stores it either in `ptr` or in `ary` (`RSTRING_EMBED_LEN_MAX` is 24 on a 64 bit platform)

```c
struct RString {
  struct RBasic basic;
  union {
    struct {
      long len;
      char *ptr;   // <----- HERE...
      union {
        long capa;
        VALUE shared;
      } aux;
    } heap;
    char ary[RSTRING_EMBED_LEN_MAX + 1];    // <----- ... OR HERE
  } as;
};
```

`RSTRING_PTR` either returns a pointer to `ary` or a copy of `ptr`.

In other words there is no guarantee, that `RSTRING_PTR` actually returns a null-terminated string. On the other hand `StringValueCStr` makes this guarantee (but most likely duplicates the string in memory) I found [this article](http://patshaughnessy.net/2012/1/4/never-create-ruby-strings-longer-than-23-characters) giving a more in-depth explanation of the implementation details.

It is obvious that - for the strings Bootsnap is working with - `RSTRING_PTR` returns a null-terminated string most (all?) of the times, but if I understand the code correctly, there are cases where the hash calculation of Bootsnap might go over the memory boundaries of the string. 

In my opinion `StringValueCStr` should be used instead or - if that leads to performance problems - `RSTRING_PTR` in combination with `RSTRING_LEN`.

I'm willing to create a PR for this, but someone with a better understanding of the inner workings of Ruby and/or C should review this. Perhaps there are other implementation details which actually do guarantee null-terminated strings and all this is just a theoretical problem...

**Comment by burke** [comment]
Interesting! (and sorry for the huge delay; I took September off)

I think I'd prefer a solution that used ptr+len. It'd be just as easy to adapt the function for that case.

I see we're calling it on cstrings from libc in a couple of cases. Maybe we should just copy the fnv1a functions since they're so simple.

**Comment by casperisfine** [comment]
42b9cd5 now use `RSTRING_LEN` for iterating over Ruby strings.
